### PR TITLE
Offroad power monitoring

### PR DESCRIPTION
**Pull request by robbederks**
Needs to be tested

### CODE REVIEW

**Comment by robbederks** [comment]
Also, waiting on commaai/cereal#24

**Comment by geohot** [comment]
Merged cereal PR

File: selfdrive/thermald/power_monitoring.py [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+  with open('/sys/class/power_supply/battery/charging_enabled', 'w') as f:
+    f.write(f"{1 if on else 0}\n")
+
+last_measurement_time = None           # Used for integration delta
+power_used_uWh = 0                    # Integrated power usage in uWh since going into offroad
+next_pulsed_measurement_time = None
+
+# Calculation tick
+def pm_calculate(health):
+  global power_used_uWh, last_measurement_time, next_pulsed_measurement_time
```
Comment (pd0wm) [inline selfdrive/thermald/power_monitoring.py]: Why not turn this into a class so you don't need globals?

Comment (robbederks) [reply selfdrive/thermald/power_monitoring.py]: Could do that. Will refactor and work on this more tomorrow

File: selfdrive/thermald/power_monitoring.py [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+def set_battery_charging(on):
+  with open('/sys/class/power_supply/battery/charging_enabled', 'w') as f:
+    f.write(f"{1 if on else 0}\n")
+
+# Helpers
+def _read_param(path, parser, default=0):
+  try:
+    with open(path) as f:
+      return parser(f.read())
+  except FileNotFoundError:
```
Comment (pd0wm) [inline selfdrive/thermald/power_monitoring.py]: What if your parser function fails to parse?

File: selfdrive/thermald/power_monitoring.py [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+
+def get_usb_present():
+  return _read_param("/sys/class/power_supply/usb/present", lambda x: bool(int(x)), False)
+
+def get_battery_charging():
+  # This does correspond with actually charging
+  return _read_param("/sys/class/power_supply/battery/charge_type", lambda x: x.strip() != "N/A", False)
+
+def set_battery_charging(on):
+  with open('/sys/class/power_supply/battery/charging_enabled', 'w') as f:
```
Comment (pd0wm) [inline selfdrive/thermald/power_monitoring.py]: handle exception?

**Pull request merged**